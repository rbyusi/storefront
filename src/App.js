import React, { useState, useEffect } from "react";
import Header from "./Components/Header";
import MediaBanner from "./Components/MediaBanner";
import Items from "./Components/Items";
import Cart from "./Components/Cart";

function App() {
  const [Products, setProducts] = useState([]);
  const [CartItems, setCartItems] = useState([]);
  const [isItemSelected, setItemSelected] = useState(false);
  const [isCartSelected, setIsCartSelected] = useState(false);
  const [ViewItem, setViewItem] = useState();

  const fetchProducts = async () => {
    console.log(process.env.NODE_ENV)
    var url =  process.env.NODE_ENV === "development" ?  "http://localhost:5000" : "https://store-front-items.herokuapp.com"
    const res = await fetch(`${url}/products`);
    const data = await res.json();
    
    return data;
    
  };

  const onViewItem = (id) => {
    const item = Products.find((item) => id === item.id);
    setItemSelected((v) => true);
    setViewItem(item);
  };

  const backToItems = () => {
    setItemSelected((v) => false);
    toggleViewCart();
  };

  const toggleViewCart = () => setIsCartSelected(!isCartSelected);

  const onAddItem = (product, bulk = 0) => {
    const itemExists = CartItems.find((item) => item.id === product.id);
    const itemCount = bulk > 0 ? bulk : 1;
    if (itemExists) {
      setCartItems(
        CartItems.map((item) =>
          item.id === product.id
            ? { ...itemExists, quantity: itemExists.quantity + itemCount }
            : item
        )
      );
    } else {
      setCartItems([...CartItems, { id: product.id, quantity: itemCount }]);
    }
  };

  const onRemoveItem = (product, remove) => {
    const itemExists = CartItems.find((item) => item.id === product.id);
    const removeItem = remove ? "yes" : "no";
    if (itemExists && removeItem === "no") {
      setCartItems(
        CartItems.map((item) =>
          item.id === product.id
            ? { ...itemExists, quantity: itemExists.quantity - 1 }
            : item
        )
      );
    } else if (removeItem === "yes") {
      setCartItems(CartItems.filter((item) => item.id !== product.id));
    }
  };

  useEffect(() => {
    const getProducts = async () => {
      const ProductsFromServer = await fetchProducts();
      setProducts(ProductsFromServer);
    };
    getProducts();
  }, []);

  return (
    <div id="Container">
      <Header
        onAddItem={onAddItem}
        onRemoveItem={onRemoveItem}
        CartItems={CartItems}
        backToItems={backToItems}
        toggleViewCart={toggleViewCart}
      />
      <MediaBanner isItemSelected={isItemSelected} ViewItem={ViewItem} isCartSelected={isCartSelected}/>
      <Items
        Products={Products}
        CartItems={CartItems}
        onAddItem={onAddItem}
        onViewItem={onViewItem}
        isItemSelected={isItemSelected}
        ViewItem={ViewItem}
        isCartSelected={isCartSelected}
      />
      <Cart
        Products={Products}
        onAddItem={onAddItem}
        onRemoveItem={onRemoveItem}
        CartItems={CartItems}
        backToItems={backToItems}
        isCartSelected={isCartSelected}
      />
    </div>
  );
}

export default App;
