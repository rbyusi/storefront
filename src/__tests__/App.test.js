import { cleanup, render, screen } from "@testing-library/react";
import renderer from 'react-test-renderer'

import App from "../App";

beforeEach(()=>{
  render(<App />);
})

afterEach(()=>{
  cleanup();
})

test("Renders the whole /app component properly",()=>{
    const app = renderer.create(<App />)
    let tree = app.toJSON();
    expect(tree).toMatchSnapshot();
})

test("Successfuly enders the navbar", () => {

  const myCart = screen.getByText("My Cart (0)");
  expect(myCart).toBeInTheDocument();
});

test("Successfuly renders the mediabanner", () => {

  const mediabanner = screen.getByText("Plates")
  expect(mediabanner).toBeInTheDocument();
});
