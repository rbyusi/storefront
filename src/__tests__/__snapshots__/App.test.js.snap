// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the whole /app component properly 1`] = `
<div
  id="Container"
>
  <div
    className="Header"
  >
    <span>
      <img
        alt="Hero"
        src="/Images/logo.png"
      />
    </span>
    <span
      className="Links"
    >
      <ul>
        <li>
          Home
        </li>
        <li
          onClick={[Function]}
        >
          Shop
        </li>
        <li>
          Journal
        </li>
        <li>
          More
        </li>
      </ul>
    </span>
    <span
      data-testid="MyCart"
      id="Cart"
      onClick={[Function]}
    >
      My Cart (0)
    </span>
  </div>
  <div
    className="Banner"
  >
    <div
      className="BannerBox"
    >
      <h1>
        Plates
      </h1>
      <p>
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
      </p>
    </div>
  </div>
  <div
    id="Items"
  />
  <div
    className="CartSelected"
  >
    <h1>
      Shopping Cart
    </h1>
    <div
      className="CartItems"
    >
      <div
        className="Headers"
      >
        <div
          className="Product"
        >
          Product
        </div>
        <div
          className="Quantity"
        >
          Quantity
        </div>
        <div
          className="Total"
        >
          Total
        </div>
        <div
          className="Action"
        >
          Action
        </div>
      </div>
      <div>
        <div
          className="ShoppingCartItems"
        >
          <div
            className="NoCartItem"
          />
        </div>
        <div
          className="CartOverView"
        />
        <div
          className="CheckOut"
        >
          <h2
            onClick={[Function]}
          >
            Continue Shopping
          </h2>
           
          <div />
        </div>
      </div>
    </div>
  </div>
</div>
`;
